core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,1099 @@
+                param_changes: [],
+                dest: 'Introduction',
+                feedback: {
+                  html: '',
+                  content_id: 'content'
+                },
+                labelled_as_correct: true,
+                refresher_exploration_id: 'exp',
+                missing_prerequisite_skill_id: null
+              },
+              confirmed_unclassified_answers: [],
+              id: null
+            }
+          }
+        }
+      },
+      version: 1,
+      can_edit: true,
+      preferred_audio_language_code: 'en',
+      preferred_language_codes: [],
+      auto_tts_enabled: true,
+      correctness_feedback_enabled: true,
+      record_playthrough_probability: 1
+    };
+  });
+
+  beforeEach(() => {
+    alertsService = TestBed.inject(AlertsService);
+    answerClassificationService = TestBed.inject(AnswerClassificationService);
+    contextService = TestBed.inject(ContextService);
+    expressionInterpolationService =
+      TestBed.inject(ExpressionInterpolationService);
+    questionObjectFactory = TestBed.inject(QuestionObjectFactory);
+    readOnlyExplorationBackendApiService = TestBed.inject(
+      ReadOnlyExplorationBackendApiService);
+    questionPlayerEngineService = TestBed.inject(QuestionPlayerEngineService);
+    outcomeObjectFactory = TestBed.inject(OutcomeObjectFactory);
+    urlService = TestBed.inject(UrlService);
+  });
+
+  it('should load questions when initialized', () => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'getExplorationId')
+      .and.returnValue('explorationId1');
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+    spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+
+    expect(questionPlayerEngineService.getQuestionCount()).toBe(0);
+
+    questionPlayerEngineService.init(
+      multipleQuestionsBackendDict, successHandler, failHandler);
+
+    expect(questionPlayerEngineService.getQuestionCount()).toBe(3);
+  });
+
+  it('should update the exploration version if it ' +
+    'is not in question player mode when initialized', fakeAsync(() => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+    let exploration = sampleExplorationDict;
+    exploration.version = 2;
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'getExplorationId').and.returnValue('explorationId1');
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(false);
+    spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+
+
+    spyOn(readOnlyExplorationBackendApiService, 'loadExplorationAsync')
+      .and.callFake((expID, version) => {
+        expect(questionPlayerEngineService.getExplorationVersion()).toBe(1);
```

Comment:
Shouldn't this expectation be placed in the outer scope, similar to the post-check? If you put it in here I have no idea when it runs.

Reply:
I think we can't put it in this case. Since both are being updated in the same function `.init()`.
The other pre-check we can add here is to expect the version to be null before initialization.

Reply:
removed this test case,